### PR TITLE
fix settler and one city challenge 

### PR DESCRIPTION
**Pull request by MindaugasRumsa51**
tested ingame

### CODE REVIEW

File: core/src/com/unciv/models/ruleset/unique/UniqueTriggerActivation.kt [hunk 1]
```diff
@@ -78,7 +78,11 @@ object UniqueTriggerActivation {
                 return true
             }
             OneTimeFreeUnitRuins -> {
-                val unit = civInfo.getEquivalentUnit(unique.params[0])
+                var unit = civInfo.getEquivalentUnit(unique.params[0])
+                if ( unit.hasUnique(UniqueType.FoundCity) && civInfo.isOneCityChallenger()) {
+                    unit = civInfo.getEquivalentUnit(Constants.worker)
```
Comment (yairm210) [inline core/src/com/unciv/models/ruleset/unique/UniqueTriggerActivation.kt:83]: Remember that in modded games they may not have a worker unit! Or they could have one with a different name!

File: core/src/com/unciv/models/ruleset/unique/UniqueTriggerActivation.kt [hunk 2]
```diff
@@ -78,7 +78,15 @@ object UniqueTriggerActivation {
                 return true
             }
             OneTimeFreeUnitRuins -> {
-                val unit = civInfo.getEquivalentUnit(unique.params[0])
+                var unit = civInfo.getEquivalentUnit(unique.params[0])
+                if ( unit.hasUnique(UniqueType.FoundCity) && civInfo.isOneCityChallenger()) {
+                    if (ruleSet.units[Constants.worker] != null) {
+                        unit = civInfo.getEquivalentUnit(Constants.worker)
```
Comment (yairm210) [inline core/src/com/unciv/models/ruleset/unique/UniqueTriggerActivation.kt:84]: You're still using hardcoded unit names, please search for units by type and unique (like you did to see if it's a settler!)

File: core/src/com/unciv/models/ruleset/unique/UniqueTriggerActivation.kt [hunk 3]
```diff
@@ -78,7 +78,23 @@ object UniqueTriggerActivation {
...
+                    var found = false
+                    for (unitForReplace in ruleSet.units.values) {
+                        if(unitForReplace.getMatchingUniques(UniqueType.BuildImprovements)
+                           .any { it.params[0] == "Land" } ) {
+
+                            unit = civInfo.getEquivalentUnit(unitForReplace.name)
+                            found = true
+                            break
+                        }
+                    }
```
Comment (yairm210) [inline core/src/com/unciv/models/ruleset/unique/UniqueTriggerActivation.kt:92]: can be simplified to replacementUnitl = ruleSet.units.values.firstOrNull{it.getMatchingUniques(UniqueType.BuildImprovements)
                           .any { it.params[0] == "Land" }} ?: return false

**Review by yairm210 — APPROVED** [review]

**Comment by MindaugasRumsa51** [comment]
question:  in code there are 2 other similar pieces.
for example "AlertPopup.kt" file have line:
captor.placeUnitNearTile(tile.position, Constants.worker)

maybe those need fix too?

**Comment by OptimizedForDensity** [comment]
Nice catch with getting settlers from ruins on OCC. Regarding the other cases of `Constants.worker`, those can be handled in a later PR, this is good for now.

**Review by OptimizedForDensity — APPROVED** [review]

**Pull request merged**